### PR TITLE
Add function `arrayNormalizedGini`

### PR DESCRIPTION
**Pull request by ucasfl**
See https://arxiv.org/abs/1912.07753 and https://www.kaggle.com/code/batzner/gini-coefficient-an-intuitive-explanation

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added function `arrayNormalizedGini`

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: tests/queries/0_stateless/03279_normalized_gini.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+-- CREATE TABLE t
+-- (
+    -- `a1` Array(Float32),
+    -- `a2` Array(UInt32)
+-- )
+-- ENGINE = MergeTree
+-- ORDER BY tuple();
+-- 
+-- INSERT INTO t VALUES ([0.9, 0.3, 0.8, 0.7], [6, 1, 0, 2]), ([0.9, 0.3, 0.8, 0.7], [6, 1, 0, 2]), ([0.9, 0.3, 0.8, 0.7], [6, 1, 0, 2]), ([0.9, 0.3, 0.8, 0.7], [6, 1, 0, 2]);
+--
```
Comment (ucasfl) [inline tests/queries/0_stateless/03279_normalized_gini.sql:18]: The test does not work due to https://github.com/ClickHouse/ClickHouse/issues/72822.

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit ae87b324cc76f343975ef64b3d8d969f64557213 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72823/ae87b324cc76f343975ef64b3d8d969f64557213/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72823/ae87b324cc76f343975ef64b3d8d969f64557213/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

**Review by rschu1ze — COMMENTED** [review]
@ucasfl Thanks a lot. I didn't look at the code yet, I had some high-level questions first.

File: docs/en/sql-reference/functions/array-functions.md [hunk 2]
```diff
@@ -3157,6 +3157,43 @@ Result:
 └─────────┘
 ```
 
+## normalizedGini
+
+Function `normalizedGini` is used to calculate the normalized Gini coefficient.
```
[the remark this answers is not in the document]
Comment (ucasfl) [reply docs/en/sql-reference/functions/array-functions.md:3162]: OK, we can rename it with `array` prefix.

[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/array-functions.md]: This [link](https://stats.stackexchange.com/a/307233) says that the normalized Gini index can be calculated from the AUC. ClickHouse already has a function `arrayAUC`. Do we really need to re-implement the normalized Gini index then?

[the remark this answers is not in the document]
Comment (ucasfl) [reply docs/en/sql-reference/functions/array-functions.md]: When calculate the model Gini coefficient, we need arrange label array by the order of predicted array, then calculate AUC.

[the remark this answers is not in the document]
Comment (ucasfl) [reply docs/en/sql-reference/functions/array-functions.md]: For example:
```sql
select arrayNormalizedGini([0.9, 0.3, 0.8, 0.7], [6, 1, 0, 2]); -- return (0.18055555555555547,0.26388888888888884,0.6842105263157893)
```
 is equivalent to

```sql
select
  pred_gini,
  label_gini,
  normalized_gini_coef
from
  (
    select
      length(ltv) as item_cnt,
      arrayMap(
        i -> array(pltv [i], ltv [i]),
        range(1, item_cnt + 1)
      ) as pairs,
      arrayReverseSort(pairs) as sorted_pairs,
      arrayReverseSort(ltv) as ltv_desc,
      arraySum(ltv) as ltv_sum,
      arrayCumSum(arrayMap(x -> x [2] / ltv_sum, sorted_pairs)) as pred_cumsum_ratio,
      arrayCumSum(arrayMap(x -> x / ltv_sum, ltv_desc)) as ltv_cumsum_ratio,
      0.5 * (item_cnt + 1) as random_gain_cumsum_ratio,
      (
        arraySum(pred_cumsum_ratio) - random_gain_cumsum_ratio
      ) / item_cnt as pred_gini,
      (
        arraySum(ltv_cumsum_ratio) - random_gain_cumsum_ratio
      ) / item_cnt as label_gini,
      pred_gini / label_gini as normalized_gini_coef
    from
      (
        select
          [0.9, 0.3, 0.8, 0.7] as pltv,
          [6, 1, 0, 2] as ltv
      )
  )
```

[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/array-functions.md]: Ok, good point :) I'll then take a closer look at the PR later today!

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
ClickHouse Integration tests (aarch64) [6/6]:
- test_refreshable_mat_view/test.py::test_real_wait_refresh[to_clause1-False-True]: #73482

**Pull request merged**